Source/WebCore/html/HTMLMediaElement.cpp

```diff
@@ -1908,6 +1905,143 @@ void HTMLMediaElement::updateActiveTextTrackCues(const MediaTime& movieTime)
+    switch (state) {
+    case SpeechSynthesisState::None:
+        setVolumeMultiplier(1);
+        m_cueBeingSpoken = nullptr;
+        if (!speechSynthesis)
+            return;
+
+        if (oldState != SpeechSynthesisState::None)
+            speechSynthesis->cancel();
+        break;
+    case SpeechSynthesisState::Speaking:
+        ASSERT(speechSynthesis);
+        setVolumeMultiplier(volumeMultiplierWhenSpeakingCueText);
+        if (speechSynthesis && oldState == SpeechSynthesisState::Paused)
+            speechSynthesis->resume();
+        else if (m_paused || m_pausedInternal)
+            pauseSpeakingCueText();
+        break;
+    case SpeechSynthesisState::Paused:
+        ASSERT(speechSynthesis);
+        if (speechSynthesis)
+            speechSynthesis->pause();
+        break;
+    }
+#else
+    UNUSED_PARAM(state);
+#endif
+}
+
+void HTMLMediaElement::speakCueText(TextTrackCue& cue)
+{
+#if ENABLE(SPEECH_SYNTHESIS)
+    if (m_cueBeingSpoken && m_cueBeingSpoken->isEqual(cue, TextTrackCue::MatchAllFields))
+        return;
+
+    ALWAYS_LOG(LOGIDENTIFIER, cue);
+
+    if (m_speechState != SpeechSynthesisState::None)
+        cancelSpeakingCueText();
+
+    m_cueBeingSpoken = &cue;
+    cue.speak(m_reportedPlaybackRate ? m_reportedPlaybackRate : m_requestedPlaybackRate, volume(), [weakThis = WeakPtr { *this }](const TextTrackCue&) {
+        ASSERT(isMainThread());
+        if (!weakThis)
+            return;
+
+        weakThis->setSpeechSynthesisState(SpeechSynthesisState::None);
+    });
+
+    if (m_pausedInternal || m_paused)
+        setSpeechSynthesisState(SpeechSynthesisState::Paused);
+    else
+        setSpeechSynthesisState(SpeechSynthesisState::Speaking);
+#else
+    UNUSED_PARAM(cue);
+#endif
+}
+
+void HTMLMediaElement::pauseSpeakingCueText()
+{
+#if ENABLE(SPEECH_SYNTHESIS)
+    if (m_speechState != SpeechSynthesisState::Speaking)
+        return;
+
+    ALWAYS_LOG(LOGIDENTIFIER);
+    setSpeechSynthesisState(SpeechSynthesisState::Paused);
+#endif
+}
+
+void HTMLMediaElement::resumeSpeakingCueText()
+{
+#if ENABLE(SPEECH_SYNTHESIS)
+    if (m_speechState != SpeechSynthesisState::Paused)
+        return;
+
+    ALWAYS_LOG(LOGIDENTIFIER);
+    setSpeechSynthesisState(SpeechSynthesisState::Speaking);
+#endif
+}
+
+void HTMLMediaElement::cancelSpeakingCueText()
+{
+#if ENABLE(SPEECH_SYNTHESIS)
+    if (m_speechState == SpeechSynthesisState::None)
+        return;
+
+    ALWAYS_LOG(LOGIDENTIFIER);
+    setSpeechSynthesisState(SpeechSynthesisState::None);
+#endif
+}
+
+bool HTMLMediaElement::shouldSpeakCueTextForTime(const MediaTime& time)
+{
+#if ENABLE(SPEECH_SYNTHESIS)
+    if (!m_cueBeingSpoken)
+        return false;
+
+    ALWAYS_LOG(LOGIDENTIFIER, "time = ", time, ", returning ", (time.toDouble() >= m_cueBeingSpoken->startTime() && time.toDouble() < m_cueBeingSpoken->endTime()));
```

Comment:
`(time.toDouble() >= m_cueBeingSpoken->startTime() && time.toDouble() < m_cueBeingSpoken->endTime())` can be its own variable like `timeIsInRange`

Reply:
Yes, I'll change that.